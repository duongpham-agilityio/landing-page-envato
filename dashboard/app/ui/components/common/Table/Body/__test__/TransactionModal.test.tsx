import { render, screen, fireEvent } from '@testing-library/react';
import { TransactionModal } from '..';
import userEvent from '@testing-library/user-event';

describe('Transaction Modal', () => {
  const mockTransaction = {
    id: '123',
    customer: {
      firstName: 'John',
      lastName: 'Doe',
      email: 'doe@gmail.com',
      avatar: '',
      address: {
        street: '123 Main St',
        state: 'CA',
        city: 'City',
        zip: 12345,
      },
    },
  };

  it('match to snapshot', () => {
    const { container } = render(
      <TransactionModal isDelete transaction={mockTransaction} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render delete confirmation correctly', async () => {
    render(<TransactionModal isDelete transaction={mockTransaction} />);

    // Ensure delete confirmation message is rendered
    expect(
      screen.getByText(/Are you sure delete the transaction with id/i),
    ).toBeInTheDocument();
  });

  it('should call onDeleteTransaction on delete button click', async () => {
    const mockOnDeleteTransaction = jest.fn();
    render(
      <TransactionModal
        isDelete
        transaction={mockTransaction}
        onDeleteTransaction={mockOnDeleteTransaction}
      />,
    );

    // Click the delete button
    fireEvent.click(screen.getByText('Delete'));

    // Ensure onDeleteTransaction is called
    expect(mockOnDeleteTransaction).toHaveBeenCalled();
  });

  it('should call onCloseModal on cancel button click', async () => {
    const mockOnCloseModal = jest.fn();
    render(
      <TransactionModal
        isDelete
        transaction={mockTransaction}
        onCloseModal={mockOnCloseModal}
      />,
    );

    // Click the cancel button
    fireEvent.click(screen.getByText('Cancel'));

    // Ensure onCloseModal is called
    expect(mockOnCloseModal).toHaveBeenCalled();
  });

  it('should call onUpdateTransaction and reset form on form submission', async () => {
    const mockOnUpdateTransaction = jest.fn();
    const mockOnCloseModal = jest.fn();

    render(
      <TransactionModal
        transaction={mockTransaction}
        onUpdateTransaction={mockOnUpdateTransaction}
        onCloseModal={mockOnCloseModal}
        isDelete={true}
      />,
    );

    // Simulate user input
    userEvent.type(
      screen.getByLabelText<HTMLInputElement>('First Name'),
      ' Updated',
    );

    // Submit the form
    fireEvent.click(screen.getByText('Save'));

    // Ensure onUpdateTransaction is called with the updated data
    expect(mockOnUpdateTransaction).toHaveBeenCalledWith({
      ...mockTransaction,
      customer: {
        ...mockTransaction.customer,
        firstName: 'John Updated',
      },
    });

    // Ensure reset is called with the updated data
    expect(mockOnCloseModal).toHaveBeenCalled();
  });
});
