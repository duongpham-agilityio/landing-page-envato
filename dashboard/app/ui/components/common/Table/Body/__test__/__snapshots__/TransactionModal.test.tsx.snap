// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Match to snapshot 1`] = `
<div>
  <form
    class="chakra-stack css-tl3ftk"
    id="update-transaction-form"
  >
    <div
      class="css-1jqq95v"
    >
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-1ebhc9e"
          for="field-:r0:"
          id="field-:r0:-label"
        >
          First Name
        </label>
        <div
          class="chakra-input__group css-169ufh8"
          data-group="true"
        >
          <input
            class="chakra-input css-vdi49a"
            id="field-:r0:"
            name="customer.firstName"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-1ebhc9e"
          for="field-:r1:"
          id="field-:r1:-label"
        >
          Last Name
        </label>
        <div
          class="chakra-input__group css-169ufh8"
          data-group="true"
        >
          <input
            class="chakra-input css-16fpzlu"
            id="field-:r1:"
            name="customer.lastName"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="css-1jqq95v"
    >
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-1ebhc9e"
          for="field-:r2:"
          id="field-:r2:-label"
        >
          Street
        </label>
        <div
          class="chakra-input__group css-169ufh8"
          data-group="true"
        >
          <input
            class="chakra-input css-vdi49a"
            id="field-:r2:"
            name="customer.address.street"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-1ebhc9e"
          for="field-:r3:"
          id="field-:r3:-label"
        >
          State
        </label>
        <div
          class="chakra-input__group css-169ufh8"
          data-group="true"
        >
          <input
            class="chakra-input css-16fpzlu"
            id="field-:r3:"
            name="customer.address.state"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="css-1jqq95v"
    >
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-1ebhc9e"
          for="field-:r4:"
          id="field-:r4:-label"
        >
          City
        </label>
        <div
          class="chakra-input__group css-169ufh8"
          data-group="true"
        >
          <input
            class="chakra-input css-vdi49a"
            id="field-:r4:"
            name="customer.address.city"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-1ebhc9e"
          for="field-:r5:"
          id="field-:r5:-label"
        >
          Zip
        </label>
        <div
          class="chakra-input__group css-169ufh8"
          data-group="true"
        >
          <input
            class="chakra-input css-16fpzlu"
            id="field-:r5:"
            name="customer.address.zip"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="css-16xzy9q"
    >
      <button
        class="chakra-button css-nz6i3s"
        disabled=""
        form="update-transaction-form"
        type="submit"
      >
        Save
      </button>
      <button
        class="chakra-button css-vxb15r"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;

exports[`Transaction Modal match to snapshot 1`] = `
<div>
  <div
    class="css-0"
  >
    <p
      class="chakra-text css-1ezsviu"
    >
      Are you sure delete the transaction with id:
      <span
        class="chakra-text css-16ximt"
      >
        123
      </span>
      ?
    </p>
    <div
      class="css-ofbpde"
    >
      <button
        class="chakra-button css-nz6i3s"
        type="button"
      >
        Delete
      </button>
      <button
        class="chakra-button css-vxb15r"
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;
